diff --git a/CMakeLists.txt b/CMakeLists.txt
new file mode 100644
index 0000000..17a6335
--- /dev/null
+++ b/CMakeLists.txt
@@ -0,0 +1,16 @@
+cmake_minimum_required(VERSION 2.6)
+project(masstree C CXX)
+
+set(CMAKE_CXX_STANDARD 11)
+set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -std=c++11 -g -O2 -include ${CMAKE_CURRENT_SOURCE_DIR}/config.h")
+
+set(kvtreeSRCS
+    json.cc
+    straccum.cc
+    string.cc
+    clp.c
+    compiler.cc
+    memdebug.cc
+    kvthread.cc
+)
+add_library(mass-tree STATIC ${kvtreeSRCS})
